### ISSUE TITLE
Unable to get 0.6.0-beta.1 to work in Node

### ISSUE THREAD

**Issue by mbostock** [kylebarron/parquet-wasm#412]
Thanks for this library! I am trying to use the latest version published to npm, 0.6.0-beta.1. It seems that the conditional exports added in #382 cause Node to import the “esm” version rather than the “node” version.

Here is what I wrote:

```js
import * as Arrow from "apache-arrow";
import * as Parquet from "parquet-wasm";

const table = Arrow.tableFromArrays({test: [42]});
process.stdout.write(Parquet.writeParquet(Parquet.Table.fromIPCStream(Arrow.tableToIPC(table, "stream"))));
```

This results in the following error, and note that it’s importing `esm/arrow1.js`:

```
…/node_modules/parquet-wasm/esm/arrow1.js:1174
            wasm.__wbindgen_add_to_stack_pointer(16);
                 ^

TypeError: Cannot read properties of undefined (reading '__wbindgen_add_to_stack_pointer')
    at Table.fromIPCStream (…/node_modules/parquet-wasm/esm/arrow1.js:1174:18)
    at …/test-parquet.js:5:57
    at ModuleJob.run (node:internal/modules/esm/module_job:217:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:316:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:66:12)
```

I tried to fix this by importing the Node entry explicitly:

```js
import * as Arrow from "apache-arrow";
import * as Parquet from "parquet-wasm/node/arrow1.js";

const table = Arrow.tableFromArrays({test: [42]});
process.stdout.write(Parquet.writeParquet(Parquet.Table.fromIPCStream(Arrow.tableToIPC(table, "stream"))));
```

However, this isn’t allowed: 😞 

```
node:internal/errors:497
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './node/arrow1.js' is not defined by "exports" in …/node_modules/parquet-wasm/package.json imported from …/test-parquet.js
    at new NodeError (node:internal/errors:406:5)
    at exportsNotFound (node:internal/modules/esm/resolve:268:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:598:9)
    at packageResolve (node:internal/modules/esm/resolve:772:14)
    at moduleResolve (node:internal/modules/esm/resolve:838:20)
    at defaultResolve (node:internal/modules/esm/resolve:1043:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:383:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:352:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:228:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:85:39) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}
```

I was able to workaround this problem by downgrading to 0.5.0 which doesn’t have the named entries.

It seems like the top-level `import` condition is taking precedence over the `node` condition…

(Related, did you intend to make 0.6.0-beta.1 the latest tag on npm? Or did you intend it to be a prerelease? If so you may want to use the next tag to encourage folks to stay on 0.5.0 until 0.6.0 is out of beta. 🙏)

**Comment by kylebarron** [comment]
Hi! Thanks for trying out the library! I was just testing the new conditional export support in 0.6.0-beta.1, so it's good to see that it needs improvement.

> This results in the following error, and note that it’s importing `esm/arrow1.js`:

To fix the error while using this esm import, you need to await the default export. E.g. https://observablehq.com/@kylebarron/geoparquet-on-the-web#readParquet. 

<img width="704" alt="image" src="https://github.com/kylebarron/parquet-wasm/assets/15164633/09569ed8-e397-43da-88f9-3f2ad07b74af">

> It seems like the top-level `import` condition is taking precedence over the `node` condition…

I'm not super familiar with intricacies of JS bundling; do you have an idea of why that is? Both `import` and `node` are listed here: https://github.com/kylebarron/parquet-wasm/blob/2204f946c6dce6ceddedd593cdf6847ff18c9c50/templates/package.json#L57-L71

> I tried to fix this by importing the Node entry explicitly:

Hmm, maybe we should add specific exports for e.g. `/node`? 

> (Related, did you intend to make 0.6.0-beta.1 the latest tag on npm? Or did you intend it to be a prerelease? If so you may want to use the next tag to encourage folks to stay on 0.5.0 until 0.6.0 is out of beta. 🙏)

Oh I didn't know that was possible. I'm less familiar with NPM distribution compared to PyPI, where everything with an alpha or beta tag is considered a pre-release. I guess I need to look at npm publish options.

I'm sick this week but I'll try to improve this usability when I'm healthy!

**Comment by mbostock** [comment]
> To fix the error while using this esm import, you need to await the default export.

Understood, but I’m in Node and I don’t want the esm import — I want the one targeting Node, which isn’t happening based on how the conditional imports are declared in the package.json.

I think that’s because Node itself supports both import and require, and in my case, I’m using type: "module" and hence Node is favoring the “import” conditional over the “node” conditional. Per the [conditional exports](https://nodejs.org/api/packages.html#conditional-exports) docs:

> Within the ["exports"](https://nodejs.org/api/packages.html#exports) object, key order is significant. During condition matching, earlier entries have higher priority and take precedence over later entries. The general rule is that conditions should be from most specific to least specific in object order.

So I think you probably need something like this but I haven’t tested…

```json
    ".": {
      "node": {
        "types": "./node/arrow1.d.ts",
        "default": "./node/arrow1.js"
      },
      "types": "./esm/arrow1.d.ts",
      "default": "./esm/arrow1.js"
    }
```

This should cause Node to use the Node-targeted entries, and everything else will use the esm bundle.

**Comment by kylebarron** [comment]
I'm unable to reproduce your original issue. With node `v20.4.0` 
```json
{
  "dependencies": {
    "parquet-wasm": "^0.6.0-beta.1"
  }
}
```
and a `tmp.mjs` of 
```js
import * as Parquet from "parquet-wasm";
console.log(Parquet)
```
that seemed to just work.

In any case, I made https://github.com/kylebarron/parquet-wasm/pull/414. Are you able to see if that fixes the issues on your end? Or would it be easier if I published a `beta.2` for you to test?

**Comment by net** [comment]
This is still an issue on beta.2.

**Comment by llimllib** [comment]
@kylebarron unfortunately node can't run #414 as far as I can tell, npm fails with `Cannot read properties of null (reading 'edgesOut')`. Repro steps:

```
$ tmp=$(mktemp -d); cd $tmp; npm init -y; npm install 'git+https://github.com/kylebarron/parquet-wasm.git#63c288fc116a5fddb3527b3302c9d6285a9cbb0b' 
Wrote to /private/var/folders/cj/3fpctnd52vv37y9gpdgg7c200000gn/T/tmp.ubiQKbhhui/package.json:

{
  "name": "tmp.ubiqkbhhui",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}


npm ERR! code 1
npm ERR! git dep preparation failed
npm ERR! command /Users/llimllib/.local/share/mise/installs/node/20.11.1/bin/node /Users/llimllib/.local/share/mise/installs/node/20.11.1/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/Users/llimllib/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! npm ERR! Cannot read properties of null (reading 'edgesOut')
npm ERR! 
npm ERR! npm ERR! A complete log of this run can be found in: /Users/llimllib/.npm/_logs/2024-03-14T12_30_09_391Z-debug-0.log

npm ERR! A complete log of this run can be found in: /Users/llimllib/.npm/_logs/2024-03-14T12_30_09_009Z-debug-0.log

$ node --version && npm --version
v20.11.1
10.2.4
```

**Comment by llimllib** [comment]
It also seems that node can't import `parquet-wasm` on the current head (24402e37) of this repo:

```
tmp=$(mktemp -d); cd $tmp; npm init -y; npm pkg set type="module"; npm install 'git+https://github.com/kylebarron/parquet-wasm.git'; echo 'import * as Parquet from "parquet-wasm"; console.log(Parquet)' >| test.js; node test.js

<snip>

npm ERR! Cannot read properties of undefined (reading 'spec')

npm ERR! A complete log of this run can be found in: /Users/llimllib/.npm/_logs/2024-03-14T12_47_14_826Z-debug-0.log
node:internal/modules/esm/resolve:853
  throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base), null);
        ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'parquet-wasm' imported from /private/var/folders/cj/3fpctnd52vv37y9gpdgg7c200000gn/T/tmp.qouJLtqIKQ/test.js
    at packageResolve (node:internal/modules/esm/resolve:853:9)
    at moduleResolve (node:internal/modules/esm/resolve:910:20)
    at defaultResolve (node:internal/modules/esm/resolve:1130:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:396:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:365:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:240:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:85:39)
    at link (node:internal/modules/esm/module_job:84:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v20.11.1
```

**Comment by llimllib** [comment]
OK, I take it all back, the sample code #414 works, it just took me a long time to figure out how. Steps:

1. Clone this repo
2. check out `kyle/fix-package-exports`
3. set up `PATH` for mac as directed in `DEVELOP.md`
4. install `wasm-pack`
5. run `scripts/build.sh`
6. replace `parquet-wasm` directory in `node_modules` with the `pkg` directory after a successful build
7. run the code that initially started this thread

**Comment by kylebarron** [comment]
I published 0.6.0 just now, and believe this to be resolved. Reopen or add a new comment if you find a bug.

**Comment by llimllib** [comment]
Thanks!
